src/java/org/apache/cassandra/utils/concurrent/ExtendedNonBlockingHashMap.java

```diff
@@ -0,0 +1,202 @@
+                    catch (Throwable t)
+                    {
+                        newEntry.setFailure(t);
+                        // Remove future so that construction can be retried later
+                        remove(key, future);
+                    }
+                }
+                else
+                {
+                    newEntry = null;
+                }
+
+                // Else some other thread beat us to it, but we now have the reference to the future which we can wait for.
+            }
+
+            try
+            {
+                future.syncUninterruptibly();
+            }
+            catch (Throwable t)
+            {
+                // if blockingUnloadIfPresent was called in the meantime, we simply retry hoping that unloading gets
+                // finished soon
+                // also we retry if the concurrent attempt to load entry failed (but we do not retry if this attempt
+                // failed)
+                if (newEntry == null || Throwables.isCausedBy(t, ex -> ex instanceof KeyNotFoundException))
+                {
+                    Thread.yield();
+                    continue;
+                }
+            }
+
+            future.rethrowIfFailed();
+            return future.getNow();
+        }
+    }
+
+    /**
+     * If a value for the given key is present, unload function is run and the value is removed from the map.
+     * Similarly to {@link #blockingLoadIfAbsent(Object, Supplier)} at-most-once semantics is guaranteed for unload
+     * function.
+     * <p>
+     * When unload function fails, the value is removed from the map anyway and the failure is rethrown.
+     * <p>
+     * When the key was not found, the method returns {@code null}.
+     *
+     * @throws UnloadExecutionException when the unloading failed to complete - this is checked exception because
+     *                                  the value is removed from the map regardless of the result of unloading; therefore if the unloading failed, the
+     *                                  called is responsible for handling that; the {@link UnloadExecutionException} encapsulates the value which was
+     *                                  failed to unload.
+     */
+    public V blockingUnloadIfPresent(K key, Consumer<? super V> unloadFunction) throws UnloadExecutionException
+    {
+        Promise<V> droppedFuture = new AsyncPromise<V>().setFailure(new KeyNotFoundException());
+
+        Future<V> existingFuture;
+        do
+        {
+            existingFuture = get(key);
+            if (existingFuture == null || existingFuture.cause() != null)
+                return null;
+        } while (!replace(key, existingFuture, droppedFuture));
+
+        V v = existingFuture.awaitUninterruptibly().getNow();
+        if (v == null)
+        {
+            // which means that either the value failed to load or a concurrent attempt to unload already did the work
+            return null;
```

Comment:
At this point we have replaced the entry with our `droppedFuture`, which will no longer be removed from the map.

Reply:
yes, thanks for this catch. I don't know how it happened. I must have changes something right before committing because even tests were not passing without fixing this :/